spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/route/TimeoutConfig.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.gateway.route;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.springframework.validation.annotation.Validated;
+
+/**
+ * Configuration for per-route http timeouts.
+ *
+ * @author Stefan Stus
+ */
+@Validated
+public class TimeoutConfig {
+
+	public static final TimeoutConfig EMPTY_TIMEOUT_CONFIG = new TimeoutConfig();
+
+	/** The connect timeout in millis */
+	private Optional<Integer> connectTimeout;
+
+	/** The response timeout. */
+	private Optional<Duration> responseTimeout;
+
+	TimeoutConfig() {
+		connectTimeout = Optional.empty();
+		responseTimeout = Optional.empty();
+	}
+
+	private TimeoutConfig(Integer connectTimeout, Duration responseTimeout) {
+		this.connectTimeout = Optional.ofNullable(connectTimeout);
+		this.responseTimeout = Optional.ofNullable(responseTimeout);
+	}
+
+	public static TimeoutConfig timeoutConfigOf(Integer connectTimeout,
+			Duration responseTimeout) {
+		return new TimeoutConfig(connectTimeout, responseTimeout);
+	}
+
+	public static TimeoutConfigBuilder builder() {
+		return new TimeoutConfigBuilder();
+	}
+
+	public Optional<Integer> getConnectTimeout() {
+		return connectTimeout;
+	}
+
+	public void setConnectTimeout(Integer connectTimeout) {
+		this.connectTimeout = Optional.ofNullable(connectTimeout);
+	}
+
+	public Optional<Duration> getResponseTimeout() {
+		return responseTimeout;
+	}
+
+	public void setResponseTimeout(Duration responseTimeout) {
+		this.responseTimeout = Optional.ofNullable(responseTimeout);
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		TimeoutConfig that = (TimeoutConfig) o;
+		return Objects.equals(connectTimeout, that.connectTimeout)
+				&& Objects.equals(responseTimeout, that.responseTimeout);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(connectTimeout, responseTimeout);
+	}
+
+	@Override
+	public String toString() {
+		return "TimeoutConfig{" + "connectTimeout=" + connectTimeout
+				+ ", responseTimeout=" + responseTimeout + '}';
+	}
+
+	public static class TimeoutConfigBuilder {
+
+		private Integer connectTimeout;
```

Comment:
it would be awesome to make this field of type `Duration`. @spencergibb what do you think? or does it make to leave it of the same type as default config?